### PR TITLE
feat: Linux AppImage packaging + auto-update support

### LINKED ISSUE #124: feat: add AppImage packaging with auto-update support for Linux
## Summary

Add AppImage as a Linux distribution format with built-in auto-update support, extending the auto-update system currently being implemented for Windows (see design spec: `docs/superpowers/specs/2026-03-16-auto-update-design.md`).

Currently Linux is planned to only show "update available" notification with a manual download link. With AppImage, we can provide true auto-update on Linux — matching the Windows experience.

## Current Auto-Update Architecture

The existing implementation uses:
- **`electron-updater`** for version checking (reads `latest.yml` from GitHub Releases)
- **Manual download** via Node.js `https` module (because Squirrel.Windows is incompatible with `electron-updater`'s `autoDownload`)
- **IPC channels**: `update-check`, `update-download`, `update-install` (renderer→main), `update-status`, `update-progress` (main→renderer)
- **UI**: `UpdateBanner` (top bar notification) + `UpdateDialog` (changelog + buttons) + Settings button
- **CI**: GitHub Actions builds artifacts → creates Draft release → developer manually publishes

Linux is currently handled as a special case in `UpdateManager`: on `update-available`, it sends a `downloadUrl` pointing to the GitHub Release page for manual download. No auto-download or auto-install.

## Why AppImage

- **Universal**: Runs on virtually all Linux distros without installation
- **Self-contained**: Bundles all dependencies, no system package conflicts
- **Auto-updatable**: `electron-updater` natively supports AppImage auto-update (unlike .deb or .zip)
- **In-place update**: Downloads new AppImage, replaces old binary, restarts — seamless UX

AppImage is the **only** Linux format where `electron-updater`'s full `autoDownload` + `quitAndInstall()` flow works out of the box.

### Current Linux Audio Architecture

The app currently uses `pulseaudio-utils.js` which directly calls system commands to manage audio:

**System audio capture** (`sokuji_system_audio_*`):
- `pactl load-module module-null-sink` — create placeholder null-sink
- `pactl load-module module-remap-source` — create remap-source visible to browser
- `pw-link` — connect sink monitor to virtual mic
- `pw-link -d` — disconnect physical mic auto-connections

**Virtual speaker/mic for TTS output** (`sokuji_virtual_output` + `sokuji_virtual_mic`):
- `pactl load-module module-null-sink` — create virtual speaker sink
- `pactl load-module module-remap-source` — create virtual mic from sink monitor
- `pw-link` — connect virtual speaker output to virtual mic input
- `pw-link -d` — disconnect physical ports from virtual mic

**Cleanup**: `pactl unload-module`, `pactl list modules short` for orphan detection

### Simplification with electron-audio-loopback

`electron-audio-loopback` already supports Linux (via `getDisplayMedia` loopback) and is already used on Windows/macOS. Extending it to Linux could **replace the system audio capture portion** (`sokuji_system_audio_*`), eliminating the need for:
- `sokuji_system_audio_null` null-sink creation
- `sokuji_system_audio_mic` remap-source creation  
- `pw-link` connections for system audio capture
- `LinuxLoopbackRecorder.ts` (which currently uses `getUserMedia` with the virtual mic device)

**However**, the virtual speaker/mic for TTS output injection (`sokuji_virtual_output` + `sokuji_virtual_mic`) still requires `pactl` and `pw-link` — this cannot be replaced by `electron-audio-loopback` since it's about creating audio output devices, not capturing.

### Key Investigation: AppImage + Host Audio Commands

**Critical question**: Can an AppImage call the host system's `pactl` and `pw-link`?

AppImage is self-contained but typically runs **unsandboxed** — it can access host binaries via `$PATH`. However, this needs verification:

- [ ] **Investigate**: Can AppImage access host `pactl` and `pw-link` commands?
- [ ] **Investigate**: Does the PulseAudio/PipeWire socket (`$XDG_RUNTIME_DIR/pulse/`, `$XDG_RUNTIME_DIR/pipewire-0`) work from within AppImage?
- [ ] **Investigate**: Do `pactl load-module` operations from AppImage properly create devices visible to the host audio system?
- [ ] **Investigate**: Are there permission or D-Bus issues when calling audio commands from AppImage context?
- [ ] **Investigate**: Does `electron-audio-loopback` work correctly on Linux when running as AppImage? (it uses Electron's `getDisplayMedia` which may need screen capture portal access)
- [ ] **Fallback plan**: If host commands are not accessible, explore alternatives:
  - Bundle `pactl`/`pw-link` inside AppImage (version compatibility risk)
  - Use PulseAudio/PipeWire client libraries directly via native Node addon
  - Use D-Bus API to communicate with PipeWire/PulseAudio directly

### Recommended Approach

1. **First**: Enable `electron-audio-loopback` on Linux to replace system audio capture (`sokuji_system_audio_*`) — reduces PulseAudio/PipeWire operations needed
2. **Then**: Investigate whether remaining `pactl`/`pw-link` calls for virtual TTS devices work from within AppImage
3. **Finally**: Build AppImage and test the full audio pipeline

### 1. AppImage Build

Add AppImage output to the build pipeline. Since Electron Forge's AppImage support is limited, the recommended approach is:

**Option A: `@electron-forge/maker-appimage`**
- Simpler if it works with our forge.config.js setup
- May have limitations with electron-updater integration

**Option B: Use `electron-builder` for Linux AppImage only**
- Mature AppImage + electron-updater integration
- Requires parallel build config for Linux alongside existing Forge config

### 2. Extend UpdateManager for AppImage

In `electron/update-manager.js`, the current Linux path just sends a download link. For AppImage:

```javascript
// Current behavior
if (process.platform === 'linux') {
  // Send downloadUrl for manual download
}

// Proposed: detect if running as AppImage
const isAppImage = !!process.env.APPIMAGE;

if (process.platform === 'linux' && isAppImage) {
  // Use electron-updater's full auto-update flow
  // electron-updater natively supports AppImage replacement
  autoUpdater.autoDownload = false; // still require user confirmation
  // On install: replaces the AppImage binary and restarts
} else if (process.platform === 'linux') {
  // Fallback for .deb/.zip users: manual download link (current behavior)
}
```

Key difference from Windows: **no Squirrel incompatibility**. `electron-updater` can handle the full AppImage update cycle natively — no manual `https` download workaround needed. `autoUpdater.quitAndInstall()` works directly.

### 3. CI/CD: Build & Publish AppImage

Update `.github/workflows/build.yml`:

- Build AppImage for Linux x64 and ARM64
- Generate `latest-linux.yml` manifest (electron-updater reads this for Linux updates)
- Upload AppImage + `latest-linux.yml` to GitHub Release alongside existing artifacts (.deb, .zip)

```yaml
# latest-linux.yml format
version: 0.16.0
files:
  - url: Sokuji-0.16.0-x86_64.AppImage
    sha512: <hash>
    size: <bytes>
path: Sokuji-0.16.0-x86_64.AppImage
sha512: <hash>
releaseDate: '2026-03-16T00:00:00.000Z'
```

### 4. No UI Changes Needed

The existing `UpdateBanner`, `UpdateDialog`, and `updateStore` already handle all states (checking, available, downloading, downloaded, error). The only change is in `UpdateManager` — AppImage users get the full download+install flow instead of the manual download link.

### Phase 1: Audio Investigation & Simplification
- [ ] Investigate: enable `electron-audio-loopback` on Linux for system audio capture
- [ ] Investigate: can AppImage access host `pactl`, `pw-link`, PulseAudio/PipeWire sockets?
- [ ] Investigate: does `electron-audio-loopback` `getDisplayMedia` work in AppImage (XDG portal access)?
- [ ] If feasible: replace `LinuxLoopbackRecorder` + `sokuji_system_audio_*` with `electron-audio-loopback`
- [ ] Test remaining virtual TTS device operations (`sokuji_virtual_output`/`sokuji_virtual_mic`) from AppImage context

### Phase 2: AppImage Packaging
- [ ] Choose AppImage build approach (Forge maker vs electron-builder)
- [ ] Add AppImage build configuration
- [ ] Update `.github/workflows/build.yml` to build AppImage for x64 and ARM64
- [ ] Test AppImage launches and audio pipeline works end-to-end

### Phase 3: Auto-Update
- [ ] Update `electron/update-manager.js`: detect AppImage via `process.env.APPIMAGE`, use native electron-updater flow
- [ ] Keep manual download link fallback for non-AppImage Linux (.deb, .zip)
- [ ] Generate `latest-linux.yml` in CI from built AppImage
- [ ] Upload AppImage + manifest to GitHub Release
- [ ] Test full update cycle: v1 AppImage → detect v2 → download → replace → restart

### Phase 4: Edge Cases
- [ ] Handle FUSE requirement (fallback: `--appimage-extract-and-run`)
- [ ] Handle `--no-sandbox` flag if needed on some distros
- [ ] Ensure file permissions after update (AppImage must remain executable)
- [ ] Test fallback: non-AppImage Linux users still get manual download link
- [ ] Keep .deb and .zip builds as alternative Linux formats

## AppImage Considerations

- **FUSE requirement**: AppImage needs FUSE to mount. Most desktop distros include it, but some minimal/container environments don't. Users can use `--appimage-extract-and-run` as fallback.
- **File size**: AppImage bundles everything (~150-200MB expected), larger than .deb. Delta updates can reduce subsequent download sizes.
- **Permissions**: After auto-update, the new AppImage must retain executable permission. `electron-updater` handles this.
- **Desktop integration**: AppImage doesn't auto-register in app menus. Tools like `AppImageLauncher` or `appimaged` handle this for users who want it.

## References

- [electron-updater AppImage support](https://www.electron.build/auto-update)
- [electron-audio-loopback](https://github.com/alectrocute/electron-audio-loopback) — supports macOS, Windows, and Linux
- [AppImage documentation](https://docs.appimage.org/)
- Auto-update design spec: `docs/superpowers/specs/2026-03-16-auto-update-design.md`
- Current auto-update implementation: `electron/update-manager.js`, `src/stores/updateStore.ts`
- Current Linux audio: `electron/pulseaudio-utils.js`, `src/lib/modern-audio/LinuxLoopbackRecorder.ts`
- Related: #107 (Flatpak packaging)

### PR DESCRIPTION
**Pull request by jiangzhuo**
Closes #124.

## Summary

Add AppImage (x64 + arm64) as the recommended Linux distribution format with native `electron-updater` auto-update. Keep `.deb` for existing users but drop `.zip`. `.deb` users now see a banner proposing migration to AppImage with two download buttons.

Linux packaging moves from Electron Forge (`maker-deb` + `maker-zip`) to `electron-builder` (AppImage + deb). Windows (Squirrel) and macOS (PKG) remain on Forge.

## Why

- `electron-updater`'s `AppImageUpdater` provides first-class in-place auto-update for AppImage, which neither `.deb` nor `.zip` supports.
- The existing "Linux shows update available + download link" flow was dormant anyway — no `latest-linux*.yml` was produced, so `electron-updater`'s check always silently 404'd. This PR finally makes Linux notifications work for all formats.
- `electron-builder` is the reference implementation for the `electron-updater` ecosystem — we stop paying Windows's Squirrel-compatibility tax on Linux.

## What changed

**Build pipeline:**
- `package.json`: added `build` field (electron-builder config, Linux AppImage + deb targets, publish provider for `latest-linux*.yml` generation); added `electron-builder` devDep; removed `@electron-forge/maker-deb` + `@electron-forge/maker-zip`.
- `scripts/electron-builder-fuses.js` (new): `afterPack` hook applying the same Electron Fuses the Forge `FusesPlugin` set.
- `forge.config.js`: Linux makers deleted; Windows/macOS makers + FusesPlugin + `packageAfterPrune` hook untouched.
- `.github/workflows/build.yml`: arm64 Linux row switched from `ubuntu-latest` (cross-compile) to native `ubuntu-24.04-arm`; Forge build step now Windows-only; new `electron-builder` step for Linux; release asset collection swaps `.zip` for `.AppImage` + `latest-linux*.yml`.

**Runtime:**
- `electron/update-manager.js`: added `this.isAppImage = !!process.env.APPIMAGE` detection; Linux `update-available` branch populates `supportsAutoUpdate`, `appImageUrl`, `debUrl`, `releasePageUrl`; `update-download`/`update-install` handlers branch on AppImage: native `autoUpdater.downloadUpdate()` + `quitAndInstall()` on AppImage, no-op on non-AppImage Linux, Windows Squirrel path preserved. Listener cleanup on error/retry.
- `src/stores/updateStore.ts`: 4 new state fields + 4 selectors + 3 unit tests (151 → 154 tests).

**UI & i18n:**
- `UpdateDialog`: three-way footer (non-AppImage Linux → two-button migration variant; legacy `downloadUrl` → Go to Download; else Download Now) + auto-update note + `.tertiary-button` style.
- `UpdateBanner`: label swaps to `linuxMigrateTitle` when `supportsAutoUpdate === false`.
- 5 new i18n keys (`autoUpdateNote`, `downloadAppImage`, `downloadDeb`, `linuxMigrateTitle`, `linuxMigrateBody`) added to all 30 locales; native localizations for zh_CN, zh_TW, ja, ko, de, fr, es, pt_BR, ru.

## Out of scope

- **Windows NSIS migration** — Forge `maker-squirrel` → `electron-builder` NSIS would unify with Linux and unlock native Windows auto-update too. Tracked as #207.
- **AppImage GPG signing** — electron-builder supports it; we have no Linux signing key yet. Keep unsigned (consistent with current `.deb`).
- **Flatpak / apt repo** — tracked separately.

## Test plan

- [x] Local build: `npx electron-builder --linux AppImage deb --x64` produces `Sokuji-0.19.0-x86_64.AppImage` (264 MB), `sokuji_0.19.0_amd64.deb` (140 MB), `latest-linux.yml` with SHA512 hashes for both.
- [x] Audio pipeline verification: real AppImage launched on Linux host; `process.env.APPIMAGE` set; `pactl` reachable from inside AppImage; virtual TTS devices (`sokuji_virtual_output` + `sokuji_virtual_mic`) created successfully; end-to-end translation session works.
- [x] Test suite: 154/154 passing (+3 new tests in `updateStore.test.ts`).
- [ ] **E2E auto-update cycle** (post-merge; requires published releases):
  - [ ] Cut `v0.20.0-rc1`, wait for CI, publish draft on GitHub
  - [ ] Install `Sokuji-0.20.0-rc1-x86_64.AppImage` on Linux host
  - [ ] Cut `v0.20.0-rc2`, publish
  - [ ] From running rc1: Check for Updates → download → "Restart and Update" → relaunches as rc2
  - [ ] Verify AppImage binary replaced in place
- [ ] **E2E .deb migration banner** (post-merge):
  - [ ] Install `sokuji_0.20.0-rc1_amd64.deb`
  - [ ] Trigger update check
  - [ ] Banner reads "New version v0.20.0-rc2 available" (migrate title)
  - [ ] Dialog shows auto-update note + Download AppImage (recommended) + Download .deb + Later
  - [ ] "Download AppImage" opens correct browser URL
  - [ ] No auto-download / no `update-downloaded` event in logs

## Design docs

- Spec: `docs/superpowers/specs/2026-04-20-linux-appimage-auto-update-design.md`
- Plan: `docs/superpowers/plans/2026-04-20-linux-appimage-auto-update.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * AppImage is the recommended Linux format with native in-app auto-update; non‑AppImage Linux users get migration UI and external download options.

* **Packaging / Build**
  * CI now builds AppImage and .deb for x64/arm64, uses an arm64 Linux runner, and uploads AppImage/deb plus update manifests.

* **UI**
  * Update banner/dialog show format-specific copy and buttons for AppImage/.deb when appropriate.

* **Localization**
  * Added translations for the Linux/AppImage update flow.

* **Documentation**
  * Added design and migration planning docs.

* **Tests**
  * Added tests for the update store fields.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,6 +10,66 @@
...
+      "artifactName": "${productName}-${version}-${arch}.${ext}"
+    },
+    "deb": {
+      "artifactName": "${name}_${version}_${arch}.${ext}"
+    },
+    "publish": {
+      "provider": "github",
+      "owner": "kizuna-ai-lab",
+      "repo": "sokuji"
+    },
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply package.json:70]: Keeping `build.publish` because it is still required for `latest-linux*.yml` generation (a `publish: null` config suppresses manifest emission, which electron-updater then cannot read). Prevented the actual publish race instead by adding `--publish never` to the CI electron-builder invocation — see fix in df55ae61 and the related comment thread on `.github/workflows/build.yml`.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -87,6 +87,19 @@ jobs:
           VITE_ENABLE_VOLCENGINE_AST2: ${{ secrets.VITE_ENABLE_VOLCENGINE_AST2 }}
           VITE_ENABLE_PALABRA_AI: ${{ secrets.VITE_ENABLE_PALABRA_AI }}
 
+      - name: Build Linux packages (electron-builder)
+        if: runner.os == 'Linux'
+        run: npx electron-builder --linux AppImage deb --${{ matrix.arch || 'x64' }}
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply .github/workflows/build.yml]: Fixed in df55ae61 — added `--publish never` to the electron-builder step. `GH_TOKEN` is kept because the `release` job downstream still needs it, and because electron-builder uses it for release metadata lookup (not for actual upload, now that `--publish never` is explicit).

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -87,6 +87,19 @@ jobs:
...
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          CI: false
+          VITE_BACKEND_URL: ${{ secrets.VITE_BACKEND_URL_PROD || 'https://sokuji-api.kizuna.ai' }}
+          VITE_ENVIRONMENT: production
+          VITE_POSTHOG_KEY: ${{ secrets.VITE_POSTHOG_KEY }}
+          VITE_ENABLE_VOLCENGINE_ST: ${{ secrets.VITE_ENABLE_VOLCENGINE_ST }}
+          VITE_ENABLE_VOLCENGINE_AST2: ${{ secrets.VITE_ENABLE_VOLCENGINE_AST2 }}
+          VITE_ENABLE_PALABRA_AI: ${{ secrets.VITE_ENABLE_PALABRA_AI }}
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply .github/workflows/build.yml:123]: Fixed in df55ae61 — `--publish never` added to the Linux electron-builder step.

File: src/stores/updateStore.test.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      downloadTotal: 0,
+      errorMessage: null,
+      downloadUrl: null,
+      supportsAutoUpdate: true,
+      appImageUrl: null,
+      debUrl: null,
+      releasePageUrl: null,
+      bannerDismissed: false,
+      dialogOpen: false,
+    });
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply src/stores/updateStore.test.ts:22]: Correct observation — `beforeEach` sets the fields explicitly so the "defaults" assertions were verifying the reset, not the true initial state. Renamed the tests in df55ae61 to reflect what they actually check (schema-level guards against field removal / type drift). True-initial-state testing against a Zustand singleton would require `vi.resetModules()` + re-import, which is more complexity than this guard warrants.

File: docs/superpowers/specs/2026-04-20-linux-appimage-auto-update-design.md [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+}
+```
+
+Notes:
+- `files` is the glob equivalent of Forge's function-based `ignore`. Exclude `build/wasm/**` then re-include runtime dirs explicitly (same semantics as current whitelist).
+- `extraResources` preserves `assets/` and `resources/` in the app resource dir.
+- Artifact names:
+  - **AppImage**: `Sokuji-${version}-${arch}.AppImage` where `${arch}` is `x64` / `arm64` (electron-builder's internal arch names for AppImage).
+  - **deb**: `sokuji_${version}_${arch}.deb` where `${arch}` is auto-mapped by electron-builder's deb target to Debian conventions (`x64` → `amd64`, `arm64` → `arm64`). UpdateManager must use the Debian-mapped value when constructing `debUrl`.
+- `publish: null` — we don't use electron-builder's publisher; GitHub Release upload stays in the existing CI workflow.
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply docs/superpowers/specs/2026-04-20-linux-appimage-auto-update-design.md]: Spec updated in df55ae61 to match the implementation: `publish` is now documented as the GitHub provider block (with reasoning — manifest generation requires a configured publisher), and the AppImage arch substitution is documented as `x86_64` (electron-builder substitutes `x86_64` in `artifactName` for x64 Linux AppImage builds, even though `linux.target.arch` config uses the `x64` name).

File: docs/superpowers/plans/2026-04-20-linux-appimage-auto-update.md [hunk 6]
```diff
@@ -0,0 +1,1220 @@
...
+    ],
+    "category": "AudioVideo",
+    "icon": "assets/icon.png",
+    "executableName": "sokuji"
+  },
+  "appImage": {
+    "artifactName": "${productName}-${version}-${arch}.${ext}"
+  },
+  "deb": {
+    "artifactName": "${name}_${version}_${arch}.${ext}"
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply docs/superpowers/plans/2026-04-20-linux-appimage-auto-update.md:139]: **Not fixing as suggested — the direction is inverted.**

The real artifact name from electron-builder 25.1.8 is `x86_64`, not `x64`:

```
$ cat out/make-linux/latest-linux.yml
version: 0.19.0
files:
  - url: Sokuji-0.19.0-x86_64.AppImage
    ...
path: Sokuji-0.19.0-x86_64.AppImage
```

electron-builder uses its own arch naming when substituting `${arch}` in `appImage.artifactName`: x64 config → `x86_64` filename, arm64 config → `arm64` filename. (The `linux.target.arch` config still uses the Node-style `x64`/`arm64` — it is the `artifactName` substitution that flips to the Linux-system convention.)

`electron/update-manager.js` uses `x86_64`, which correctly matches the real artifact. If we changed it to `x64` (as the suggestion proposes), the AppImageUpdater URL would 404 and auto-update would silently break.

Fixed the other direction in df55ae61: updated the 7 stale `x64.AppImage` references in the plan file to `x86_64.AppImage` so the documentation matches the code and the actual build output.
